### ISSUE TITLE
Keep 'GoogleMapV2' branch close to 'master' branch

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#7509]
Open question:
How can we achieve that the diff between  `master` and `GoogleMapsV2` is minimized. Is there a clever way to do this, or will it require manual merging and/or cherry-picking on regular intervals?

I tried to open a simple PR for this assuming, that there should not be so many conflicts but at least initally there are quite some.  AFAICS the current implementation of GmapsV2 is reusing same files, right? Does this complicate the merging?

**Comment by moving-bits** [comment]
IMHO it's a good idea to keep the branches in sync, otherwise there is a good chance, that too much get's lost when doing huge merges.

Merging should be done in short intervals, to keep the delta (and potential conflicts) small.

Initially we probably need to do the merging more or less one by one (cherry picking) and resolve the conflicts along the way.

**Comment by moving-bits** [comment]
A different approach could be to use V2 as a separate map provider parallel to the existing ones. This would make keeping it up to date easier and broaden the testing user base.

The v2 implementation is in a separate dir & class hierarchy, so no conflicts here. Need to check though, if any of the common files are changed in a conflicting way. Also need to check the technical prerequisites first.

**Comment by Lineflyer** [comment]
For some reason we said in our virtual team meeting, that a parallel usage would not be possible so we went for the dedicated branch.

Maybe this was only due to the required targetAPI raise by end of the year....but if it works together until then, I support the idea of having both maps available.

**Comment by moving-bits** [comment]
I've done a directory comparision between current master and GoogleMapsV2 branch. Most of the differences are clear (either self explaining or attributable to some of the ongoings on master branch), and the map implementation itself is in a separate directory.

BUT: There are also a couple files in the maps directory, which are used by the other map implementations as well, not only adding new methods, but in some parts also changing the interfaces. Changes I do not understand due to lack of knowledge. So at least for me it wouldn't be possible to adapt v2 to work in parallel with v1.

I have not checked the technical prerequisites. There might be another obstacle...

**Comment by UniQP** [comment]
The general approach is to regularly rebase the branch on `master`.

**Comment by Lineflyer** [comment]
> The general approach is to regularly rebase the branch on master.

As I am no git expert:
What is the effort behind? Is this the same as a merging from `master` to `GoogleMapsV2` or somehow easier? Could you do that on regular basis?

**Comment by Narrat** [comment]
Simply said, with a rebase every commit, which isn't found in the branch you want to rebase to, is stashed. Then the branch is fast forwarded. After that the new commits are applied on top of the branch again.
Reasons for doing this? It is a cleaner history and it can be somewhat easier to handle.
How it is done? In the best case with a simple `git rebase master` which rebases your currently checked out branch with master.

**Comment by Narrat** [comment]
What could be worth reading: https://www.atlassian.com/git/tutorials/merging-vs-rebasing

**Comment by UniQP** [comment]
@Lineflyer Yes, I can regularly rebase the branch.

@rsudev It's not obvious to me how to resolve the conflict in `CGeoMap.java`. Can you rebase the branch to `bab7d29e80bfb438d9ae3996e471b14a241ddf11` and resolve the conflict?

**Comment by moving-bits** [comment]
I had tried rebasing, but failed at CGeoMap.java as well, as I did not understand the underlaying changes made in the GoogleMapsV2 branch for this (and some related) files. The other four conflicting files are easy to fix.

Resolving those conflicts are also a prerequisite if we want to try to get both map versions in parallel (if technically possible at all), which would be nice to get a broader testing base, and to keep it close to master branch.

**Comment by UniQP** [comment]
I resolved the conflicts and rebased `GoogleMapsV2` on `master`.

**Comment by Lineflyer** [comment]
@UniQP thanks for that.
However now the nightly builds for this branch are failing:

<http://ci.cgeo.org/job/cgeo%20nightly%20GMapsV2/ws/main/src/cgeo/geocaching/maps/CGeoMap.java>:791: error: incompatible types: PositionAndHistory cannot be converted to GeneralOverlay
                mapView.repaintRequired(overlayPositionAndScale);

See also the respective CI job for more details.

**Comment by UniQP** [comment]
@Lineflyer I fixed the problem and the CI seems happy: http://ci.cgeo.org/job/cgeo%20continuous%20integration/2156/

**Comment by Lineflyer** [comment]
Thanks, nightly could be built now.

How shall we continue with alligning these branches. Open another PR like the last one and merge from time to time. Or will you (or someone else) go with rebasing from time to time. I have no idea what is the better choice for a clean git SCM.

**Comment by UniQP** [comment]
I can rebase it regularly.

**Comment by Lineflyer** [comment]
@UniQP Thanks. I added your mail adress to the GMV2 nightly build. So you will receive a mail if the job fails. Hope, thats fine for you, otherwise feel free to remove it on CI.

**Comment by Lineflyer** [comment]
No longer needed as we now merged GMapsV2 to master.
IMHO we could also delete that `GmapsV2` branch
